eskipfile/remote.go

```diff
@@ -0,0 +1,146 @@
+package eskipfile
+
+import (
+	"errors"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/routing"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type remoteEskipFile struct {
+	preloaded       bool
+	remotePath      string
+	localPath       string
+	eskipFileClient *WatchClient
+	threshold       int
+	verbose         bool
+}
+
+// RemoteWatch creates a route configuration client with (remote) file watching. Watch doesn't follow file system nodes,
+// it always reads (or re-downloads) from the file identified by the initially provided file name.
+func RemoteWatch(rf string, th int, v bool) (routing.DataClient, error) {
+	if !isFileRemote(rf) {
+		return Watch(rf), nil
+	}
+
+	tempFilename, err := ioutil.TempFile("", "routes")
+
+	if err != nil {
+		return nil, err
+	}
+
+	dataClient := &remoteEskipFile{
+		remotePath: rf,
+		localPath:  tempFilename.Name(),
+		threshold:  th,
+		verbose:    v,
+	}
+
+	err = dataClient.DownloadRemoteFile()
+
+	if err != nil {
+		return nil, err
```

Comment:
I would defer the first download until the first call to LoadAll. The reason is that there can always be temporary hickups, and there's no reason for preventing Skipper to start up on the failure of the source.

What makes this a bit complex: is the default 404 OK in such cases? Our usual approach is that we don't report healthy from Skipper until the first successful route fetch.

Reply:
I see from the usage of the `preloaded` field that this is a conscious decision. Can you give some rationale for this approach?

Reply:
I think all of this is a side effect of us trying to work around https://github.com/zalando/skipper/issues/1709 not working.

What we were doing (using skipper as a library) was to call force `DownloadRemoteFile` when the dataClient was created, and then when `Skipper.run` was being invoked, the file was already downloaded, and startup is faster (eskip files can be several MB, and download is not instantaneous).

But with #1709 being fixed now, I'm not sure all this logic is still needed.

Reply:
I added a draft of how it would look like.